salt/modules/keystoneng.py

```diff
@@ -0,0 +1,382 @@
+import salt.utils
+
+HAS_SHADE = False
+try:
+    import shade
+    from shade.exc import OpenStackCloudException
+    HAS_SHADE = True
+except ImportError:
+    pass
+
+
+def __virtual__():
+    if HAS_SHADE:
+        return 'keystoneng'
+    return False
+
+
+__opts__ = {}
+
+
+def compare_changes(obj, **kwargs):
+    changes = {}
+    for k, v in obj.items():
+        if k in kwargs:
+            if v != kwargs[k]:
+                changes[k] = kwargs[k]
+    return changes
+
+
+def get_entity(cloud, ent_type, **kwargs):
+    try:
+        func = 'keystoneng.{}_get'.format(ent_type)
+        ent = __salt__[func](cloud=cloud, **kwargs)
+    except OpenStackCloudException as e:
+        # NOTE(SamYaple): If this error was something other than Forbidden we
+        # reraise the issue since we are not prepared to handle it
+        if 'HTTP 403' not in e.inner_exception[1][0]:
+            raise
+
+        # NOTE(SamYaple): The user may not be authorized to perform the action
+        # they are trying to do, but not authorized to search. In such a
+        # situation we want to trust that the user has passed a valid ID, even
+        # though we cannot validate that this is a valid id
+        ent = kwargs['name']
+
+    return ent
+
+
+def _clean_kwargs(keep_name=False, **kwargs):
+    if 'name' in kwargs and not keep_name:
+        kwargs['name_or_id'] = kwargs.pop('name')
+
+    try:
+        clean_func = salt.utils.args.clean_kwargs
+    except AttributeError:
+        clean_func = salt.utils.clean_kwargs
+    return clean_func(**kwargs)
```

Comment:
This looks good so that the module could be dropped into `_modules` in older versions of salt :+1:

Reply:
yea im working with an older version which is why i added this